### PR TITLE
Added LossFunctionWrapper for losses and simplified code

### LINKED ISSUE #1324: Make LossFunctionWrapper public
**Describe the feature and the current behavior/state.**
We just wrapper loss function by ourselves and public ```LossFunctionWrapper``` will unify the behavior.
**Relevant information**
- Are you willing to contribute it (yes/no): Yes
- Are you willing to maintain it going forward? (yes/no): yes
- Is there a relevant academic paper? (if so, where):
- Is there already an implementation in another framework? (if so, where):Yes
- Was it part of tf.contrib? (if so, where):

**Which API type would this fall under (layer, metric, optimizer, etc.)**
Losses
**Who will benefit with this feature?**
Developers.
**Any other info.**
For example:
```
@tf.keras.utils.register_keras_serializable(package="Addons")
class GIoULoss(LossFunctionWrapper):
    """Implements the GIoU loss function.

    GIoU loss was first introduced in the
    [Generalized Intersection over Union:
    A Metric and A Loss for Bounding Box Regression]
    (https://giou.stanford.edu/GIoU.pdf).
    GIoU is an enhancement for models which use IoU in object detection.

    Usage:

    ```python
    gl = tfa.losses.GIoULoss()
    boxes1 = tf.constant([[4.0, 3.0, 7.0, 5.0], [5.0, 6.0, 10.0, 7.0]])
    boxes2 = tf.constant([[3.0, 4.0, 6.0, 8.0], [14.0, 14.0, 15.0, 15.0]])
    loss = gl(boxes1, boxes2)
    print('Loss: ', loss.numpy())  # Loss: [1.07500000298023224, 1.9333333373069763]
    ```
    Usage with tf.keras API:

    ```python
    model = tf.keras.Model(inputs, outputs)
    model.compile('sgd', loss=tfa.losses.GIoULoss())
    ```

    Args:
      mode: one of ['giou', 'iou'], decided to calculate GIoU or IoU loss.
    """

    @typechecked
    def __init__(
        self,
        mode: str = "giou",
        reduction: str = tf.keras.losses.Reduction.AUTO,
        name: Optional[str] = "giou_loss",
    ):
        super(GIoULoss, self).__init__(
            giou_loss, name=name, reduction=reduction, mode=mode)
```

### PR DESCRIPTION
**Pull request by ashutosh1919**
Fixes #1324 .  
The ideas of the fix are taken from PR #1329 .  
@gabrieldemarmiesse , please review.

### CODE REVIEW

**Comment by ashutosh1919** [comment]
@AakashKumarNain , @gabrieldemarmiesse and @WindQAQ - If the code added for `LossFunctionWrapper` in this PR is correct, then it is exposing some of the wrong test cases written for some loss in my opinion. Please give your comments.

**Comment by ashutosh1919** [comment]
@WindQAQ , I have undone changes for npairs and sparsemax_loss. I think it is good to create separate issue to correct logical errors in these two losses and then we can include wrapper to it.

**Comment by ashutosh1919** [comment]
@WindQAQ , let me also dive deeper into `npairs` and `sparsemax_loss`. Let's change that too.

File: tools/testing/source_code_test.py [hunk 1]
```diff
@@ -91,6 +91,11 @@ def test_no_private_tf_api():
         "tensorflow_addons/optimizers/moving_average.py",
         "tensorflow_addons/metrics/r_square.py",
         "tensorflow_addons/utils/test_utils.py",
+        "tensorflow_addons/losses/contrastive.py",
+        "tensorflow_addons/losses/focal_loss.py",
+        "tensorflow_addons/losses/lifted.py",
+        "tensorflow_addons/losses/quantiles.py",
+        "tensorflow_addons/losses/triplet.py",
```
Comment (gabrieldemarmiesse) [inline tools/testing/source_code_test.py:98]: Friendly reminder that those rules are here to enforce the usage of the public TF API only, by contributors in the future, so that we can be forward compatible with tensorflow. The blacklist is here to keep track of the things we have not migrated yet, and so should not grow :) Maybe I should have written a clearer message in the comment above.
